### ISSUE TITLE
Proposal: Dynamic API for PAYMENT_VARIANTS

### ISSUE THREAD

**Issue by Natureshadow** [jazzband/django-payments#298]
Currently, the available payment variants need to be configured in the DJango project settings, statically.

I want to propose a replacement for this, which introduces a dynamic API to replace this.

* Add a function `get_payment_variant(variant, payment=None)`
* Make the default implementation lookup the `variant` in `settings.PAYMENT_VARIANTS`, so it is fully backwards-compatible
* Introduce a setting `PAYMENT_VARIANT_CALLBACK`, which takes a dotted path to a function, so projects can override the function

This is useful in some scenarios, including the following two which I am facing in the [AlekSIS](https://aleksis.org)` project:

* Projects that store settings in the database, or in some way along these lines want to provide a configuration UI for payment providers instead of ahrd-coding in `settings`
* Projects with multi-tenant capabilities, where different payments might be linked to different tenants, that have different provider configuration

If this appeals to the maintainers, I would like to take on making this change.

**Comment by WhyNotHugo** [comment]
This has been discussed before (I can't find the issue right now), and it's a desirable feature. It's actually something I'd find useful too.

> Add a function get_payment_variant(variant, payment=None)

That would be `payments.core.provider_factory` right now. It takes a variant name and returns a provider instance.

> Introduce a setting PAYMENT_VARIANT_CALLBACK, which takes a dotted path to a function, so projects can override the function

`PAYMENT_VARIANT_FACTORY`, but your idea is quite on point. Function signature would be the same as the existing one.

> If this appeals to the maintainers, I would like to take on making this change.

That would be quite welcome. Adding the `PAYMENT_VARIANT_FACTORY` setting, documenting it and defaulting it to the current factory would be a great start.

The next step would be figuring out the database schema / models. I think a model with: name(str), active(bool), type(str), settings(json) should work. Users who don't use it will have an empty table, so this might need a bit more discussion.

**Comment by Natureshadow** [comment]
> The next step would be figuring out the database schema / models. I think a model with: name(str), active(bool), type(str), settings(json) should work. Users who don't use it will have an empty table, so this might need a bit more discussion.

I'd leave that as an exercise to whoever wants to use the feature, just like with the `Payment` model. The default implementation does not need a database model, and my dynamic scenario here doesn't need one either.

So, whoever wants something else than the settings dictionary should override the factory, and if their approach happens to be database-backed, then it's their task to come up with a model in their app.

Thus, I'd limit this feature to adding the setting.

**Comment by WhyNotHugo** [comment]
Yup, that second step can be done later. It might make sense to even be a separate library.

Making the factory configurable is already a pretty good improvement.
